Initial HTTP API (#2)
nexus-rpc/api: SPEC.md

Reviewer: Is this repo ever going to be more than just this file? If so, do we really need an entire repo? Maybe a repo for "specs"?

Author: I created this a while ago and never used it, I could see a case for merging the two repos.

https://github.com/nexus-rpc/designs

Can call it specs if you prefer.

Author: I'll likely rename _this_ repo if we go the direction you suggested.
I'm still wanting a tool for code generation and we may want to add some CI here to at least verify the JSON schemas make sense and importable (as a submodule?) from other repos so we could do some code gen.

Author: Let's leave this as a dedicated repo, it may grow and get JSON schema or other IDLs or an openapi definition.

# Nexus RPC HTTP Specification

## Overview

> The Nexus protocol, as specified below, is a synchronous RPC protocol. Arbitrary length operations are modelled on top

Reviewer 2: I know that the term has been in use for a while, but I personally vote for "arbitrary duration operations" rather than "length". It's time that's being referred to, so duration is more correct. (If an acronym starts getting thrown around, then it's interesting that "ADO" brings "async do" to mind -- possibly fortuitous?)

Author: @rylandg please consider

Reviewer: ADO makes me think of my old MS coding days and Active Data Objects (but I don't disagree...there's no TLA that isn't already taken somehow)

Reviewer 3: IIRC, ALO was not intended to be the published term. Both ALO and ADO feel long/complex to me. Would rather prioritize approachability/learnability over precision/accuracy with something shorter & more familiar like "async operation" or better yet, "operation".

Like maybe: 
- headline: Nexus is a standard [or spec?] for async APIs. (technically incomplete, but effectively differentiates it from other API specs)
- term used everywhere: operation
- in docs in one place: In Nexus, an **operation** is an RPC of arbitrary duration. A caller makes an HTTP request to start an operation, and in the response, the server may provide the result or a handle to the operation (which may complete and have a result at a future time).

Reviewer: You need something to differentiate a regular operation and an operation that returns a "promise" if you will. Async operation is fine with me. I do not believe the HTTP API proposal is where to hash out naming (also, consult with Ryland here).

of a set of pre-defined synchronous RPCs.

A Nexus **caller** calls a **handler**. The handler may respond inline or return a reference for a future, asynchronous
operation. The caller can cancel an asynchronous operation, check for its outcome, or fetch its current state. The caller
can also specify a callback URL, which the handler uses to asynchronously deliver the result of an operation when it
is ready.

## Operation Addressability

An operation is addressed using three components:

- [Service Name](#service-name)
- [Operation Name](#operation-name)
- [Operation ID](#operation-id)

> **Note:** More details regarding IDs will be provided upon finalizing the ID proposal.

### Service Name

Service names:

- MUST not be empty.
- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].

> 

Reviewer 2: Will the handler be case sensitive wrt service and operation name?

Reviewer: Yes, everything should be case sensitive, like URLs are

Valid characters include:

```
ALPHA / DIGIT / "-" / "." / "_" / "~"
```

### Operation Name

Operation names:

- MUST not be empty.
- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].

Valid characters include:

```
ALPHA / DIGIT / "-" / "." / "_" / "~"
```

### Operation ID

Operation IDs:

- MUST not be empty.
- MUST consist of only unreserved URL characters as delineated in [RFC3986 Section 2.3][rfc3986-section-2.3].

Valid characters include:

```
ALPHA / DIGIT / "-" / "." / "_" / "~"
```

## Schema Definitions

### Failure

> The `Failure` object represents outcomes that are `failed` or `canceled`. The object MUST adhere to the following JSON

Reviewer 2: Are "failed" and "canceled" precisely defined in this context? If not then perhaps double quotes or no quotes rather than backticks here.

schema:

> ```yaml

Reviewer: This needs a "type" or "kind" or something with some predefined values for failed, canceled, and not-done-yet.

Author: I was wondering about that, the only time we pass this around it should be accompanied by a `Nexus-Operation-State` header.

Reviewer: But state is not the same as failure type/kind. You can have a running state but a failure kind of "not-done-yet" if the wait timed out the way I am thinking about it.

Author: Are you also talking about application failure types or just framework level failure types?

Reviewer: The failure can be either. In fact, aside from state, we should use this failure object for every single failure we have (including not found, 5xx, etc).

I can be convinced either way here, but we need some failure for not-done-yet when requesting and wait times out (or no wait) while still running.

Author: I'm leaning towards using 204 with a `Nexus-Operation-State: running` header as shown below (and I may require that successful completion is returned with a 200 code).

I will, however, use this failure object for all non successful HTTP responses.

Author: Having a hard time justifying adding a type field here.
This will always come in either application failure context with the proper header or request failure accompanied by an HTTP status code.

type: object
properties:
  message:
    type: string
  details:
    type: object

>     properties:

Reviewer: I don't think you need to define the body of the details, it should be any arbitrary object that comes back from the handler. If you want metadata, that should move to a top-level field, not under details IMO

Author: This is basically a Temporal `Payload` with slight alterations, I think it makes sense this way.

Reviewer: And double-encoding of arbitrary "data" is hard to read and causes us lots of annoyance. At least https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto has `Any` "details". We should have "any details" (can be collection if needed). But I guess double encoding isn't the end of the world, just really ugly to see JSON embedded into another JSON field with escaped quotes and what not.

Author: I think we're going to want more of a convention here to interpret the details.

Reviewer 4: I agree with Chad. If we want to have metadata be an interprable thing, why not have it just be at the top? IE: Why not just have `message`, `metadata`, and `data` be the top three fields? What is the benefit of nesting the latter two inside `details`?

Author: >  just really ugly to see JSON embedded into another JSON field with escaped quotes and what not.

Agree, but we need some type of specification here so we can rehydrate these failures into language objects.

I'm wondering how often we'll see plain JSON vs. encrypted content here in practice.

So to understand the proposal from your end, you want the following, right?

```
{
  type: string (let's discuss this field separately above)
  message: string
  metadata: string -> string
  details: any (can also name this data)
}
```

So then details is any JSONable and metadata may be used to describe what goes in there...
The only problem I see is that we will want to translate this into a Temporal payload and we would need some convention to know whether details is a string that encodes binary data or should be interpreted as a plain string.
So maybe reserve a metadata field to express that?

Author: Another option would be to have separate mutually exclusive fields for binary vs. non binary data.

Reviewer 4: I like that option. Binary data can say how it is encoded in the metadata, and non-binary data can just be json. Does it even need to be separate fields? I mean I suppose _technically_ you could have raw binary that happens to also be valid json but that seems so unlikely as to be unimportant.

Author: I think the idea here is that JSON is not encoded into a string, it would be inlined.
Then if details is a string, it may represent either a plain string or some encoded data.
We need to know whether to treat it as a string or decode it.

Reviewer: > I think the idea here is that JSON is not encoded into a string, it would be inlined.

:+1: Same as an actual string, hence the "any" not "any object". But I say don't worry about binary. It should _always_ be some JSON primitive, and if they want binary, they encode (we'd have to anyways).

Author: I just want this spec to include the method by which we tell binary data encoded as string from plain string.

I'm going with

```
{
  "metadata": {
     "Content-Transfer-Encoding": "base64",
  },
  "details": "<base64-encoded-string>",
}
```

So if details is a string and metadata["Content-Transfer-Encoding"] is `base64` we'll consider details as bytes.

Author: I just realized that message may be encoded too.

New proposal:

```
{
  message: string | Payload // may be a payload
  details: any // may be a Payload
}
```

Where `Payload` is:
```
{
  metadata: map[string]string
  data: base64 encoded string
}
```

And when transferred (either via request or response) headers should include:
```
-- Option 1 --
Content-Type: application/nexus-payload-failure+json
-- Option 2 --
Nexus-Failure-Encoding: payloads | none
```

Author: Thought about it some more and encrypting the message can be done by encrypting the entire body and providing the appropriate http headers.

Author: Now debating whether the framework should support binary details.

Reviewer: Just keep it basic "any JSON" and let the user figure out how to represent binary since it's not a JSON type (so base64 string is probably most common for them, but protocol shouldn't care).

      metadata:
        type: object
        additionalProperties:
          type: string
        description: |
          String to string mapping, may contain information for decoding the data field.
      data:
        type: string
        description: |
          Arbitrary string-encoded data; use encodings like base64 for binary data transmission.
```

> ### OperationStarted

Reviewer: I think operation started and operation info should be the same thing. The result to a `POST /api/v1/services/{service}/operations/{operation}/{operation_id}` and `GET /api/v1/services/{service}/operations/{operation}/{operation_id}` should be the same structure if we can.

Author: 200 for post returns the result of the operation, not the info.
OperationStarted is only returned when an operation is started asynchronously (201), and the state of the operation is always running at that point.

Reviewer: Sorry, I don't mean change anything about 200. I mean 201 should return the same `OperationInfo` as the other call. So `OperationStarted` result and `OperationInfo` result should be the same object IMO. If that means both need the operation ID on the object and this result will always return "running", that works for me.

Author: Yeah that makes sense.

An `OperationStarted` object indicates the commencement of an asynchronous operation. This object MUST follow the given
JSON schema:

```yaml
type: object
properties:
  operationId:
    type: string
    description: |
      An identifier for referencing this operation. If provided in the start request, it is echoed back.

>   callbackUrlSupported:

Reviewer: I think the start should fail if the handler does not support callbacks. No need to tell the user we silently ignored it. If we need another high level way to query this about a Nexus server we can discuss it, but usually a client has a way it wants its results, it should not have to adjust at runtime (or worse cancel the call) if the preferred result delivery mechanism is not available.

Author: I can see arguments to go either way here, maybe we need an argument in the start call to list the desired server capabilities for getting the result (poll and or callback).

Reviewer: I would be ok saying, at least at first, all Nexus servers have to support callbacks for async operations. And we can relax that requirement later if we must.

Author: I'm proposing two options.

Request header:
`Nexus-Accept-Completion: callback | poll` header (can have multiple values)

Request query param (can find a better name):
`accept_completions=callback+poll`

If server does not support any of the completion methods, fail the request with `412 Precondition Failed` (works better with a header).

In response, replace `callbackUrlSupported` with `supportedCompletionMethods: ['callback', 'poll']`.

Reviewer: I don't think we should have optionals or a choice of whether a server supports something at this time. I think all servers need to both respond to the `GET .../result` (which of course can be called over and over) _and_ they all need to send to a callback URL if it's provided in the start call.

Author: SGTM, we can add this later if we see that implementors struggle to support both.

Reviewer 2: `callbackUrlSupported` doesn't feel like quite the right name: it's really saying "was a callback URL specified and if so was it accepted?". So it's a property of this client-server interaction, whereas "supported" feels like a property of the handler. I suggest something like `callbackUrlSet` or `calllbackUrlAccepted`.

    type: boolean
    description: |
      If the request specified a callback_url and the handler supports callbacks for this operation, this flag will be
      set.
      It is up to the caller to decided how to get the outcome of this call in case the handler does not support
      callbacks.
```

### OperationInfo

> The `OperationInfo` object, retrieved from the get operation info API, MUST adhere to the given schema:

Reviewer 2: Add link to Get Operation Info endpoint

```yaml
type: object
properties:
  state:
    enum:
      - succeeded

>       - failed

Reviewer 2: Does this include both application-level failures (e.g. operation involves an RPC but all retries failed) and framework-level failures that the framework caught? If so I think it might be helpful for the spec to say so explicitly.

Reviewer: This is operation state, there is no difference between application and framework level. An operation is failed or it is not. How a handler represents/relays operation failure is beyond the scope of the API spec.

>       - canceled

Reviewer 2: Should we distinguish between canceled-by-caller and canceled-by-framework?

Reviewer: No, this is the state of an operation. How handlers choose to handle _cancellation requests_ and then implement cancellation is beyond the scope of this spec.

      - running
    description: |

>       Describes the current state of an operation.

Reviewer 4: Is this purposefully un-typed? Should it be string?

```

## Endpoint Descriptions

### Start Operation

> Start an arbirary length operation.

Reviewer 2: ```suggestion
Start an arbitrary length operation.
```

The response of the operation may be delivered synchronously (inline), or asynchronously, via a provided callback or the
[Get Operation Result](#get-operation-result) endpoint.

#### Paths

> - `/api/v1/services/{service}/operations/{operation}`

Reviewer: ```suggestion
- `/{service}/{operation}`
```

Arguably the versioning and the prefixes don't matter (and/or can be server specific). And `services` and `operations` doesn't really matter if we can just reserve all `nexus-` prefixed services for potential future API use (i.e. _everything_ is a service and operation, even Nexus's own APIs if needed).

Author: This is what I had originally.
I buy removing the prefix, not sure about getting rid of (both) services and operations, it hurts extensibility.

Reviewer: > it hurts extensibility.

Can by operation ID have slashes? If so, then any extensibility need we can just say "that's a built in operation".

Author: We might even want to take it to an extreme and make "service" a Temporal specific thing, the service is essentially the entire URL up to the operation.

Reviewer 4: I think `/{service}/{operation}[/{sub-operation} ...]` makes sense

Reviewer: @bergundy - If you want to upend it and make this fairly HTTP/URI oriented, yes we can have operation just be a potentially-containing-slashes name, so long as you can disambiguate the ID at the end which you should be able to.

So yes, if you want to take it to its extreme, everything is `/{operation}` and `operation` can be `/machines/gumballMachine/twist`. Unsure how far you want to go. I just don't think we need versioning or too much separation of concepts. Not sure we need to go too far to the extreme.

I don't have a strong opinion here. Reminder to others: this is a Nexus platform protocol API, we still expect users may expose HTTP for HTTP callers that then uses the protocol API, but you can design how your HTTP API there looks more.

Author: Hmm.. you can't have slashes in the operation name if the operation ID is optional.

Author: But I think changing this whole spec to say a Nexus endpoint may contain a URL prefix and there's no concept of a service in non-Temporal Nexus makes total sense.

Then all the endpoints become:
```
{http-prefix}/{operation}/{operation_id}
```

Reviewer: Hrmm, there is likely value in a generic service abstraction/layer same as there is for gRPC. You will have to come up with some name for that kind of layer when making the outbound Nexus call from a workflow because there is a server mapping that keys off of it. And remember, that outbound Nexus call needs to support non-Temporal one day.

Author: True,

We can map a service to a URL for Temporal.

> - `/api/v1/services/{service}/operations/{operation}/{operation_id}`

Reviewer 5: Is the operation_id always known to the client when calling Start? I saw your comment about handler generated ids, this is probably affected...

Reviewer: This is in the process of changing. There are internal disagreements here, but suffice to say we're probably removing the ability for a client to identify an operation at first. I disagree with it (as I do for anything that makes things harder for callers such as having to store an identifier on a result to reference an operation), but oh well.

#### Method

`POST`

#### Query Parameters

> - `callback_url`: Optional. If the operation is asynchronous and the handler supports callbacks, it should invoke this URL once the operation's result is available.

Reviewer 3: ```suggestion
- `callback_url`: Optional. If the operation is asynchronous and the handler supports [callbacks](#callback-urls), it should invoke this URL once the operation's result is available.
```

#### Request Body

The body may contain arbitrary data. Headers should specify content type and encoding.

#### Response Codes

- `200 OK`: Operation completed successfully.

  **Headers**:

  - `Nexus-Operation-State: succeeded`

  **Body**: Arbitrary data conveying the operation's result. Headers should specify content type and encoding.

- `204 No Content`: Operation completed successfully with an empty result.

  **Headers**:

  - `Nexus-Operation-State: succeeded`

  **Body**: Empty.

- `482 Operation Failed`: Operation completed as `failed` or `canceled`.

  **Headers**:

  - `Content-Type: application/json`
  - `Nexus-Operation-State: failed | canceled`

  **Body**: A JSON serialized [`Failure`](#failure) object.

> - `201 Created`: Operation was started an will completed asynchronously.

Reviewer 3: ```suggestion
- `201 Created`: Operation was started and will be completed asynchronously.
```

  **Headers**:

  - `Content-Type: application/json`

  **Body**: A JSON serialized [`OperationStarted`](#operationstarted) object.

### Cancel Operation

Request to cancel an operation.
The operation may later complete as canceled or any other outcome.
Handlers should ignore multiple cancelations of the same operation and return successfully if cancelation was already
requested.

#### Method

`POST`

> #### Paths

Reviewer: You are inconsistent on whether you put method first or paths first in these sections

Author: Thanks for pointing it out.

- `/api/v1/services/{service}/operations/{operation}/{operation_id}/cancel`

#### Response Codes

- `202 Accepted`: Cancelation request accepted.

  **Body**: Unspecified.

- `404 Not Found`: Operation ID not recognized or references deleted.

  **Body**: Unspecified.

### Get Operation Result

Retrieve operation result.

#### Method

`GET`

#### Paths

- `/api/v1/services/{service}/operations/{operation}/{operation_id}/result`

#### Query Parameters

> - `wait_deadline`: Optional. ISO 8601 date indicating the waiting period for a result.

Reviewer: Hrmm, a fixed date is a bit rough. While it can make the URL seem more stable, it can be a bit harder to reason about compared to a duration. And TZ differences come into play, client-vs-server clock skew on short deadlines, etc. I think a timeout is much more reasonable here.

Just `wait` that accepts a momentjs/go-duration/etc-like time (e.g. num + short unit) I think is good. Specifically num + "ms", "s", "m", and "h" is supported (it's not that hard for servers to support and clear to users).

Author: I was debating this myself, ended up with deadline, but after reading some more I agree with you and will switch back to duration.

  If not specified, and the operation is still running, the request should resolve immediately with the current

>   operation status.

Reviewer: I think it should fail with a not-done-yet. The only success result to this call should ever be a success result with the result.

Author: I get the part about success being reserved for completion, not sure about using failure for `running` state.

Reviewer: But it is a failure of getting a response, not a failure of the operation. Like a 404. Can use different status code if you want (though not 408, my reading is showing me that has certain HTTP connection lifetime expectations).

#### Response Codes

- `200 OK`: Operation completed successfully.

  **Headers**:

  - `Nexus-Operation-State: succeeded`

  **Body**: Arbitrary data conveying the operation's result. Headers should specify content type and encoding.

> - `204 No Content`: Operation completed successfully with an empty result or is still running.

Reviewer: This is a special case where I think we should not return 204 even if the handler returned no data. I should be able to == 200 this call and know the body is what was sent even with no body. Whether there's a body or _from the handler_ I don't think affects the protocol/status in this case.

Author: I'd be fine with that, I was debating this and ended up with 204 but don't have a strong preference.

Reviewer: Yeah, I don't really have a good way to explain why 204 shouldn't be used in _all_ no-content situations. But basically in a situation where the protocol is unaware of the body size, it should just be 200 whether 1 byte body or 0 byte body. 204 is a protocol thing (i.e. "I _know_ it will be no body").

Author: I couldn't find anywhere that says you should prefer 204 over 200 for empty responses.
I could definitely change to 200, if we want to also support 204 we can add that later but I wouldn't encourage it for simplicity.

  **Headers**:

  - `Nexus-Operation-State: succeeded | running`

  **Body**: Empty.

- `482 Operation Failed`: Operation completed as `failed` or `canceled`.

  **Headers**:

  - `Content-Type: application/json`
  - `Nexus-Operation-State: failed | canceled`

  **Body**: A JSON serialized [`Failure`](#failure) object.

> - `408 Request Timeout`: The specified `wait_deadline` expired prior to operation completion.

Reviewer: After some reading, we should not use a 408 here as there's an expectation that this means connection-close should follow from the server and it has meaning to clients and servers already. Let's just reuse 482, but set state as running and the failure object's kind/type/whatever as not-done-yet.

Author: I see...
Let me consider your suggestion.

Reviewer: :+1: Yeah, reading MDN's description at https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/408 scared me with how clients, proxies, and others may make assumptions when they see this code.

Author: Same, went and read that after your first comment.

  The caller may re-issue this request to start a new long poll.

  **Body**: Unspecified.

- `404 Not Found`: Operation ID not recognized or references deleted.

  **Body**: Unspecified.

### Get Operation Info

Retrieve operation details.

#### Method

`GET`

#### Paths

- `/api/v1/services/{service}/operations/{operation}/{operation_id}`

#### Response Codes

- `200 OK`: Successfully retrieved info.

  **Headers**:

  - `Content-Type: application/json`

>   - `Nexus-Operation-State: failed | canceled`

Reviewer: ```suggestion
  - `Nexus-Operation-State: failed | canceled | running | succeeded`
```

Author: Hmm.. I think I accidentally copied this from the other endpoints.

I'm not sure we actually want this header since we're returning an OperationInfo in the body.

Reviewer: I like the consistency, but your choice

  **Body**:

  A JSON serialized [`OperationInfo`](#operationinfo) object.

- `404 Not Found`: Operation ID not recognized or references deleted.

## General Information on HTTP Response Codes

The Nexus protocol follows standard HTTP practices:

- `4xx` codes signify non-retryable framework-level errors.

> - `5xx` codes signify retryable framework-level errors.

Reviewer: Not following what framework-level means here. Unsure if we want to clarify that "retryable" in this sense is "protocol retryable". Whether a handler considers its failure retryable does not affect status codes. 4xx is non-protocol failures, 5xx is protocol failure.

Author: I'll clarify.

> - `5xx` codes signify retryable framework-level errors.

Reviewer 2: There aren't any 5xx responses in the spec.

Reviewer: It's unreasonable to define internal errors in this high level spec IMO. It may be implementation specific.

Reviewer 2: We don't need to define them in detail, but the document needs to mention that they may be returned if it's going to include this line stating that the protocol uses them with these semantics.

Reviewer: :+1: I'm sure @bergundy can add a "all calls may return 5xx errors" if it isn't otherwise obvious

## Callback URLs

Any HTTP URL can be used to deliver operation completions.

Callers should ensure URLs contain sufficient information to correlate completions with initiators.

For invoking a callback URL:

> 

Reviewer 5: You probably want to make clear that this is best effort, and is not retried forever until it returns a 200 OK
Also, it may be obvious that it uses a POST, but we could be explicit...

Reviewer 2: I wonder whether restricting it to `POST` is unnecessarily limiting. Some clients might want to use `PUT` or `PATCH`, and I can also imagine some callers being in an unfortunate position where they are only able to specify a `GET` endpoint (due to something about their server framework, privileges, etc). Might we even consider other delivery methods? I've certainly implemented things in the past that had to be delivered by email on completion of a long-running operation.

Reviewer: This is an intentionally limited API. It is the Nexus "protocol" we require servers to implement to talk to each other. A better more flexible HTTP API can be made on top (as can other protocol support). We don't want to support lots of loose implementations, that ambiguity can harm interoperability.

Author: I'll explicitly specify that we expect POST requests

- Include the `Nexus-Operation-State` header.
- If state is `succeeded`, deliver non-empty results in the body with corresponding `Content-*` headers.
- If state is `failed` or `canceled`, content type should be `application/json` and the body must have a serialized [`Failure`](#failure) object.

> - Upon successful completion delivery, the handler should reply with a `200 OK` status and an empty body.

Reviewer: Later we may have to discuss the nuances of timeouts/failures/retries when issuing callbacks

Author: Yup, that's not part of this spec though.

> - Upon successful completion delivery, the handler should reply with a `200 OK` status and an empty body.

Reviewer 3: ```suggestion
- Upon successful completion delivery, the callback recipient should reply with a `200 OK` status and an empty body.
```

to differentiate from a *Nexus handler*

### Security

> There's no enforced security for callback URLs.

Reviewer: ```suggestion
There's no enforced security for callback URLs at this time. However, some specific Nexus server implementations may deliver additional details as headers or have other security requirements of the callback endpoint.
```

Just to make it clear that a specific Nexus server can impose constraints on callback URLs.

> There's no enforced security for callback URLs.

Reviewer 2: I think some callers might be in a position where they need auth tokens set in headers for the callback request; how can they specify the required headers?

Reviewer: Not currently specified as part of the spec and we're encouraging tokens in the URL. However, specific implementations such as the Temporal server may support per endpoint/service configuration for outbound Nexus calls, like headers. Auth can get complex quick (what if they want mTLS? what if they want server-oauth? etc) and is often implementation specific, so we want to avoid a spec requiring something specific at this time.

Reviewer 2: > so we want to avoid a spec requiring something specific at this time.

I'm not suggesting requiring something -- it's the opposite; I'm suggesting making it more flexible by allowing them to supply headers required for auth.

> When starting an operation, callers may embed a signed token into the URL, which can be verified upon delivery of completion.

Reviewer 5: Do we ignore URL fragments in the provided URL? (they are supposed to be local) If so, the token should not be in the fragment...

Reviewer: URL fragments by their nature should never be sent to server, so yes, we should disallow callback URLs from even having them. Can explicitly document I guess if not an obvious implementation detail.

[rfc3986-section-2.3]: https://datatracker.ietf.org/doc/html/rfc3986#section-2.3

> ### Q/A

Reviewer 3: ```suggestion
## Q/A
```

1. What is the purpose of the `482 Operation Failed` response code? This is not a standard HTTP response code.

In internal discussions, we determined that there's value in having a specific status code for denoting operation
failures.
We chose a 4xx code because it is not meant to be retried and can be used by standard HTTP libraries and tools to
easier distinguish failure vs. success.

> Non of the standard HTTP codes fit and we decided to define a custom status code.

Reviewer 2: I'm sure everyone's aware of this but I note that the HTTP spec ([RFC9110](https://datatracker.ietf.org/doc/html/rfc9110#name-client-error-4xx)) does not say 4xx should not be retried; it just says that 4xx is a client error. So, I think it would be good to:

(a) Make sure it's clear that we're not implying that the spec says anything about whether clients should/should not retry.

(b) Justify in what sense it is a client error. If it's an operation failure (e.g. "Your operation involves socket I/O. We did all the retries you asked for and they all failed") then I think it kind of can be justified as a client error: "Your retry policy was inadequate". (But a framework error obviously isn't a client error by definition.)

```suggestion
None of the standard HTTP codes fit and we decided to define a custom status code.
```

Reviewer: To sum up the thoughts on retry, we are basically saying "you can write your Nexus client to automatically retry any 5xx errors". This is a Nexus rule we're making to help client authors, not necessarily rooted in general HTTP spec anywhere.

If it's a 4xx, whether you want to retry is a logic question, we're just saying 5xx is safe for implicit-auto-retry logic.

Reviewer 2: Since we're including this explanatory Q/A section, it might be worth explicitly saying something like

> Technically this is an abuse of 4xx, because an operation-level failure (for example because a required 3rd-party service was down) is not a client error. An alternative would be a 2xx, since the handler successfully drove the requested operation to a valid terminal state. However, we want to avoid requiring callers to inspect headers to determine whether a 2xx response contains their desired result or a `Failure` object.

2. What potential security concerns should be taken into consideration while implementing this protocol?

Security is not part of this specification, but as this is a thin layer on top of HTTP, standard practices should be

> used to secure these APIs.

Reviewer 3: ```suggestion
used to secure these APIs. For securing callback URLs, see [Callback URLs > Security](#security).
```
